src/allreduce_base.cc

```diff
@@ -241,11 +233,7 @@ utils::TCPSocket AllreduceBase::ConnectTracker(void) const {
         utils::Socket::Error("Connect");
       } else {
         fprintf(stderr, "retry connect to ip(retry time %d): [%s]\n", retry, tracker_uri.c_str());
-#if defined(_MSC_VER) || defined (__MINGW32__)
-        Sleep(retry << 1);
-#else
-        sleep(retry << 1);
-#endif
+        sleep((1 << retry) + rank%10);
```

Comment:
I don't think it resolves problem, if we have 1000 workers, still 1, 11, 21, 31....101, 111, 121....901, 911..., 991 fire connection at the same time

Reply:
expotential backoff doesn't work like this at all, 

retry 1 : (2^1)
retry 2: choose among (2^1, 2^2)
retry 3: choose among (2^1, 2^2, 2^3)

it is where randomness comes from

Reply:
>I don't think it resolves problem, if we have 1000 workers, still 1, 11, 21, 31....101, 111, 121....901, 911..., 991 fire connection at the same time

yes, this reduce concurrent reconnect from 1000 to 100 every second, assuming largest cluster is less than 2k and it seems fine for tracker to handle 200 hosts job running and reconnect in production.

> expotential backoff doesn't work like this at all

Yes, the problem is exponential back off by definition is random choice.

Reply:
we should leverage the randomness in exponential backoff instead of inventing by ourselves to avoid congestion network

Reply:
removed from this pr